pkg/storage/stores/shipper/bloomshipper/block_downloader.go

```diff
@@ -0,0 +1,233 @@
+	return &BlockDownloadingTask{
+		ctx:       ctx,
+		block:     block,
+		ErrCh:     errCh,
+		ResultsCh: resCh,
+	}
+}
+
+// noop implementation
+var onWorkerStopCallback = func() {}
+
+func (d *blockDownloader) serveDownloadingTasks(ctx context.Context, workerID string) {
+	logger := log.With(d.logger, "worker", workerID)
+	level.Debug(logger).Log("msg", "starting worker")
+
+	d.queue.RegisterConsumerConnection(workerID)
+	defer d.queue.UnregisterConsumerConnection(workerID)
+
+	idx := queue.StartIndexWithLocalQueue
+
+	for {
+		item, newIdx, err := d.queue.Dequeue(ctx, idx, workerID)
+		if err != nil {
+			if !errors.Is(err, queue.ErrStopped) && !errors.Is(err, context.Canceled) {
+				level.Error(logger).Log("msg", "failed to dequeue task", "err", err)
+				continue
+			}
+			level.Info(logger).Log("msg", "stopping worker")
+			//this callback is used only in the tests to assert that worker is stopped
+			onWorkerStopCallback()
+			return
+		}
+		task, ok := item.(*BlockDownloadingTask)
+		if !ok {
+			level.Error(logger).Log("msg", "failed to cast to BlockDownloadingTask", "item", fmt.Sprintf("%+v", item), "type", fmt.Sprintf("%T", item))
+			continue
+		}
+
+		idx = newIdx
+		blockPath := task.block.BlockPath
+		//todo add cache before downloading
+		level.Debug(logger).Log("msg", "start downloading the block", "block", blockPath)
+		block, err := d.blockClient.GetBlock(task.ctx, task.block)
+		if err != nil {
+			level.Error(logger).Log("msg", "error downloading the block", "block", blockPath, "err", err)
+			task.ErrCh <- fmt.Errorf("error downloading the block %s : %w", blockPath, err)
+			continue
+		}
+		directory, err := d.extractBlock(&block, time.Now())
+		if err != nil {
+			level.Error(logger).Log("msg", "error extracting the block", "block", blockPath, "err", err)
+			task.ErrCh <- fmt.Errorf("error extracting the block %s : %w", blockPath, err)
+			continue
+		}
+		level.Debug(d.logger).Log("msg", "block has been downloaded and extracted", "block", task.block.BlockPath, "directory", directory)
+		blockQuerier := d.createBlockQuerier(directory)
+		task.ResultsCh <- blockWithQuerier{
+			BlockRef:     task.block,
+			BlockQuerier: blockQuerier,
+		}
+	}
+}
+
+type BlockDownloadingResult struct {
+	extractedBlockPath string
+	block              BlockRef
+}
+
+func (d *blockDownloader) downloadBlocks(ctx context.Context, tenantID string, references []BlockRef) (chan blockWithQuerier, chan error) {
+	d.activeUsersService.UpdateUserTimestamp(tenantID, time.Now())
+	// we need to have errCh with size that can keep max count of errors to prevent the case when
+	// the queue worker reported the error to this channel before the current goroutine
+	// and this goroutine will go to the deadlock because it won't be able to report an error
+	// because nothing reads this channel at this moment.
+	errCh := make(chan error, len(references))
```

Comment:
Not sure if I understand that correctly. On  line 168 you return after sending the first and only error to the `errCh`. Could you just make buffered channel of size `1`?

Reply:
this channel is also passed to the queue and queue workers will report errors to this channel also...